src/lib/input/input-container.ts

```diff
@@ -232,24 +226,51 @@ export class MdInputDirective {
   constructor(private _elementRef: ElementRef,
               private _renderer: Renderer2,
               private _platform: Platform,
+              private _changeDetectorRef: ChangeDetectorRef,
               @Optional() @Self() public _ngControl: NgControl,
               @Optional() private _parentForm: NgForm,
               @Optional() private _parentFormGroup: FormGroupDirective,
               @Optional() @Inject(MD_ERROR_GLOBAL_OPTIONS) errorOptions: ErrorOptions) {
 
     // Force setter to be called in case id was not specified.
     this.id = this.id;
-
     this._errorOptions = errorOptions ? errorOptions : {};
     this.errorStateMatcher = this._errorOptions.errorStateMatcher || defaultErrorStateMatcher;
   }
 
-  /** Focuses the input element. */
-  focus() { this._elementRef.nativeElement.focus(); }
+  ngOnChanges() {
+    this._stateChanges.next();
+  }
 
-  _onFocus() { this.focused = true; }
+  ngOnDestroy() {
+    this._stateChanges.complete();
+  }
 
-  _onBlur() { this.focused = false; }
+  ngDoCheck() {
+    if (this._ngControl) {
```

Comment:
Ah very cool, this is what I was talking about when I said forms doesn't always give us a way to know when we need to update, but this is a cool workaround :)